File: day_07/Machine.hs
module Machine (
    parseLine,
    Machine(..),
    MachineOutput(..),
    evaluate,
    printOutput,
    op,
    opcode_1,
    opcode_2,
    opcode_3,
    opcode_4,
    opcode_5,
    opcode_6,
    opcode_7,
    opcode_8,
    opcode_99,
    replaceNth
) where

import Data.Array.IO
import Debug.Trace
-- import Data.List.Split

data Machine = Machine {
    values :: [Int],
    inputs :: [Int],
    outputs :: [Int],
    pos :: Int
} deriving (Show)

split :: String -> [String]
split [] = [""]
split (c:cs) | c == ',' = "" : rest
             | otherwise = (c : head rest) : tail rest
    where rest = split cs

parseLine :: String -> [Int]
parseLine "" = []
parseLine line = map read (split line)

data MachineOutput = MachineOutput {
    output :: [Int],
    resulting_opcode :: Int
} deriving (Show)

opcode :: Machine -> Int
opcode m = ((values m)!!(pos m))

op :: Machine -> Int
op m = (mod (opcode m) 100)

position :: Machine -> Int -> Int
position m x = ((values m)!!x)

immediate :: Machine -> Int -> Int
immediate m x = x

mode :: Machine -> Int -> Int
mode m 1 = (mod (div (opcode m) 100) 10)
mode m 2 = (mod (div (opcode m) 1000) 10)
mode m 3 = (mod (div (opcode m) 10000) 10)

param :: Machine -> Int -> Int
param m idx = ((values m)!!((pos m) + idx))

resolve :: Machine -> Int -> Int
resolve m idx = case (mode m idx) of
    0 -> position m (param m idx)
    1 -> immediate m (param m idx)

replaceNth :: Int -> a -> [a] -> [a]
replaceNth _ _ [] = []
replaceNth n newVal (x:xs)
    | n == 0 = newVal:xs
    | otherwise = x:replaceNth (n-1) newVal xs

opcode_1_values :: Machine -> [Int]
opcode_1_values m = replaceNth (param m 3) ((resolve m 1) + (resolve m 2)) (values m)

opcode_1 :: Machine -> Machine
opcode_1 m = (Machine (opcode_1_values m) (inputs m) (outputs m) ((pos m) + 4))

opcode_2_values :: Machine -> [Int]
opcode_2_values m = replaceNth (param m 3) ((resolve m 1) * (resolve m 2)) (values m)

opcode_2 :: Machine -> Machine
opcode_2 m = (Machine (opcode_2_values m) (inputs m) (outputs m) ((pos m) + 4))

opcode_3_values :: Machine -> [Int]
opcode_3_values m = replaceNth (param m 1) ((inputs m)!!0) (values m)

opcode_3 :: Machine -> Machine
opcode_3 m = (Machine (opcode_3_values m) (drop 1 (inputs m)) (outputs m) ((pos m) + 2))

opcode_4 :: Machine -> Machine
opcode_4 m = (Machine (values m) (inputs m) ((outputs m) ++ [resolve m 1]) ((pos m) + 2))

opcode_5_pos :: Machine -> Int
opcode_5_pos m = case (resolve m 1) of
    0 -> (pos m) + 3
    _ -> resolve m 2

opcode_5 :: Machine -> Machine
opcode_5 m = (Machine (values m) (inputs m) (outputs m) (opcode_5_pos m))

opcode_6_pos :: Machine -> Int
opcode_6_pos m = case (resolve m 1) of
    0 -> resolve m 2
    _ -> (pos m) + 3

opcode_6 :: Machine -> Machine
opcode_6 m = (Machine (values m) (inputs m) (outputs m) (opcode_6_pos m))

opcode_7_values :: Machine -> [Int]
opcode_7_values m = replaceNth
    (param m 3)
    (if (resolve m 1) < (resolve m 2) then 1 else 0)
    (values m)

opcode_7 :: Machine -> Machine
opcode_7 m = (Machine (opcode_7_values m) (inputs m) (outputs m) ((pos m) + 4))

opcode_8_values :: Machine -> [Int]
opcode_8_values m = replaceNth
    (param m 3)
    (if ((resolve m 1) == (resolve m 2)) then 1 else 0)
    (values m)

opcode_8 :: Machine -> Machine
opcode_8 m = (Machine (opcode_8_values m) (inputs m) (outputs m) ((pos m) + 4))

opcode_99 :: Machine -> MachineOutput
opcode_99 m = MachineOutput (outputs m) ((values m)!!0)

evaluate :: Machine -> MachineOutput
evaluate machine = (case (op machine) of
        1 -> evaluate (opcode_1 machine)
        2 -> evaluate (opcode_2 machine)
        3 -> evaluate (opcode_3 machine)
        4 -> evaluate (opcode_4 machine)
        5 -> evaluate (opcode_5 machine)
        6 -> evaluate (opcode_6 machine)
        7 -> evaluate (opcode_7 machine)
        8 -> evaluate (opcode_8 machine)
        99 -> opcode_99 machine
        _ -> MachineOutput [] 0)


printOutput :: Int -> IO ()
printOutput x = putStrLn ("OUTPUT: " ++ (show x))

main = do
    line <- getLine
    let values = parseLine line :: [Int]
    line <- getLine
    let inputs = parseLine line :: [Int]
    let machine = Machine values inputs [] 0
    let result = (evaluate machine)
    mapM_ printOutput (output result)
    putStrLn ("DONE " ++ (show (resulting_opcode result)))

File: day_09/Day09/Day09/Machine.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Numerics;

namespace Day09
{
    class Machine
    {
        public BigInteger Position { get; set; }
        public BigInteger RelativeIndex { get; set; }
        public Dictionary<BigInteger, BigInteger> Memory { get; set; }
        public Queue<BigInteger> Inputs;
        public delegate void OutputF(BigInteger value);
        public OutputF Output { get; set; }

        public Machine()
        {
            Position = 0;
            RelativeIndex = 0;
            Inputs = new Queue<BigInteger>();
            Output = (x) => System.Console.WriteLine("OUTPUT: " + x);
            Memory = new Dictionary<BigInteger, BigInteger>();
        }

        public Machine(String values, String inputs) : this()
        {
            BigInteger[] tmp = Array.ConvertAll(values.Split(","), BigInteger.Parse);
            for (int i = 0; i < tmp.Length; ++i)
            {
                Memory[i] = tmp[i];
            }
            if (inputs.Length > 0)
            {
                Inputs = new Queue<BigInteger>(Array.ConvertAll(inputs.Split(","), BigInteger.Parse));
            }
        }

        public void Evaluate()
        {
            while (EvaluateNext())
            {
            }
        }

        public bool EvaluateNext()
        {
            switch (op())
            {
                case 1: return Opcode1();
                case 2: return Opcode2();
                case 3: return Opcode3();
                case 4: return Opcode4();
                case 5: return Opcode5();
                case 6: return Opcode6();
                case 7: return Opcode7();
                case 8: return Opcode8();
                case 9: return Opcode9();
                case 99: return Opcode99();
                default:
                    throw new Exception("Not implemented opcode " + op().ToString());
            }
        }

        public int op()
        {
            return (int)(Memory[Position] % 100);
        }

        public bool Opcode1()
        {
            Write(3, Resolve(1) + Resolve(2));
            Position += 4;
            return true;
        }

        public bool Opcode2()
        {
            Write(3, Resolve(1) * Resolve(2));
            Position += 4;
            return true;
        }

        public bool Opcode3()
        {
            Write(1, Inputs.Dequeue());
            Position += 2;
            return true;
        }

        public bool Opcode4()
        {
            Output(Resolve(1));
            Position += 2;
            return true;
        }

        public bool Opcode5()
        {
            if (Resolve(1) != 0)
            {
                Position = Resolve(2);
            }
            else
            {
                Position += 3;
            }
            return true;
        }

        public bool Opcode6()
        {
            if (Resolve(1) == 0)
            {
                Position = Resolve(2);
            }
            else
            {
                Position += 3;
            }
            return true;
        }

        public bool Opcode7()
        {
            if (Resolve(1) < Resolve(2))
            {
               Write(3, 1);
            }
            else
            {
                Write(3, 0);
            }
            Position += 4;
            return true;
        }

        public bool Opcode8()
        {
            if (Resolve(1) == Resolve(2))
            {
                Write(3, 1);
            }
            else
            {
                Write(3, 0);
            }
            Position += 4;
            return true;
        }

        public bool Opcode9()
        {
            RelativeIndex += Resolve(1);
            Position += 2;
            return true;
        }

        public bool Opcode99()
        {
            Console.WriteLine("DONE " + Memory[0]);
            return false;
        }

        private BigInteger Resolve(int index)
        {
            BigInteger opcode = Memory[Position];
            switch (mode(opcode, index))
            {
                case 0: return Positional(index);
                case 1: return Immediate(index);
                case 2: return Relative(index);
                default:
                    break;
            }
            throw new Exception("Faulty opcode " + opcode.ToString());
        }

        private BigInteger Positional(int index)
        {
            return Read(Memory[Position + index]);
        }

        private BigInteger Immediate(int index)
        {
            return Read(Position + index);
        }

        private BigInteger Relative(int index)
        {
            return Read(RelativeIndex + Memory[Position + index]);
        }

        private BigInteger Read(BigInteger x)
        {
            BigInteger ret;
            if (Memory.TryGetValue(x, out ret))
            {
                return ret;
            }
            return 0;
        }

        private void Write(int index, BigInteger value)
        {
            Memory[GetWriteIndex(index)] = value;
        }

        private BigInteger GetWriteIndex(int index)
        {
            switch (mode(Memory[Position], index))
            {
                case 0: return Read(Position + index);
                case 2: return Read(Position + index) + RelativeIndex;
                default:
                    throw new Exception("Unexpected write mode " + Memory[Position].ToString());
            }
        }

        private int mode(BigInteger opcode, int index)
        {
            if (index == 1)
            {
                return (int)(opcode / 100 % 10);
            }
            if (index == 2)
            {
                return (int)(opcode / 1000 % 10);
            }
            if (index == 3)
            {
                return (int)(opcode / 10000 % 10);
            }
            return (int)(opcode / 100000 % 10);
        }
    }
}
